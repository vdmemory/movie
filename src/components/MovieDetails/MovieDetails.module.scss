.details {
    padding: 15px;
    width: 100%;
    max-width: 1200px;
    margin: 0 auto;
}

.iconButton {
    margin-right: 15px;
}