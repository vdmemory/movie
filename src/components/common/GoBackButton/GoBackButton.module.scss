.goBack {
    padding: 15px;
    cursor: pointer;
    background: aliceblue;
    text-align: center;
    display: flex;
    justify-content: center;
    align-items: center;
}