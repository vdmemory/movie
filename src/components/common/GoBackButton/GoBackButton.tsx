import React, { useCallback } from "react";
import { useHistory } from "react-router-dom";

import styles from "./GoBackButton.module.scss";

const GoBackButton: React.FC = ({ children }) => {
    const history = useHistory();

    const handleGoBack = useCallback(() => {
        history.goBack();
    }, [history]);

    return (
        <div className={styles.goBack} onClick={handleGoBack}>
            {children || `Back to list`}
        </div>
    );
};

export default GoBackButton;
