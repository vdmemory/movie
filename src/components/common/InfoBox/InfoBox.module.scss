.infoBox {
    border-radius: 6px;
    box-shadow: 0 0 0.2em rgba(0, 0, 0, .5);
    & .solid {
        height: 1px;
        background-color: #d3d2d2;
        width: 97%;
        margin: 0 auto;
    }
    & .title{
        font-size: 22px;
        line-height: 22px;
        padding: 15px 30px 15px;
        margin: 0;
    }
    & .content{
        & .leftBar{
            & img{}
        }
        & .rightBar{
           
        }
    }
    & .description{}
}