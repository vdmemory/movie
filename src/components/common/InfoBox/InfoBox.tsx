import React from "react";

import RowInfo from "../RowInfo/RowInfo";

import styles from "./InfoBox.module.scss";

type Props = {
    title: string;
    year: string;
    runtime: string;
    genre: string;
    director: string;
    actors: string;
    plot: string;
    language: string;
    country: string;
    poster: string;
    imdbRating: string;
};

type RowInfo = {
    title: string;
    subtitle: string;
};

const InfoBox: React.FC<Props> = ({
    title,
    year,
    poster,
    genre,
    country,
    runtime,
    director,
    actors,
    language,
    plot,
}) => {
    const infoData = [
        { title: "Original title", subtitle: title },
        { title: "Genre", subtitle: genre },
        { title: "Country", subtitle: country },
        { title: "Year", subtitle: year },
        { title: "Duration", subtitle: runtime },
        { title: "Director", subtitle: director },
        { title: "Ð¡ast of actors", subtitle: actors },
        { title: "Language", subtitle: language },
    ];

    // const renderRowInfo = ({title, subtitle}) => {
    //     return <RowInfo title={title} subtitle={subtitle} />;
    // };

    return (
        <div className={styles.infoBox}>
            <h2 className={styles.title}>{`${title} (${year})`}</h2>
            <div className={styles.solid}></div>
            <div className={styles.content}>
                <div className={styles.leftBar}>
                    <img src={poster} alt={title} />
                </div>
                <div className={styles.rightBar}>
                    {/* {infoData.map(renderRowInfo)} */}
                    right bar
                </div>
            </div>
            <div className={styles.solid}></div>
            <p className={styles.description}>{plot}</p>
        </div>
    );
};

export default InfoBox;
