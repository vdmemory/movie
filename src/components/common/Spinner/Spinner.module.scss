.wrapSpinner {
    width: 100%;
    height: 20vh;
    display: flex;
    justify-content: center;
    align-items: center;

    & .spinner {
        width: 2rem;
        height: 2rem;
        border: 3px solid transparent;
        border-top-color: rgba(black, .5);
        border-bottom-color: rgba(black, .5);
        border-radius: 50%;
        animation: rotation .8s ease infinite;
    }
}

@keyframes rotation {
    from {
        transform: rotate(0deg);
    }

    to {
        transform: rotate(360deg);
    }
}