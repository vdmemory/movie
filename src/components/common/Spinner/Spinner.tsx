import React from "react";

import styles from "./Spinner.module.scss";

const Spinner: React.FC = () => (
    <div className={styles.wrapSpinner}>
        <div className={styles.spinner}></div>
    </div>
); 
export default Spinner;
