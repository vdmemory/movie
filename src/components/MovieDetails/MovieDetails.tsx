import React, { useEffect } from "react";
import { useParams } from "react-router-dom";
import { useDispatch, useSelector } from "react-redux";

import { selectDetailsMovie } from "../../store/selectors";
import { getDetailsMovieRequest } from "../../store/actions";
import GoBackButton from "../common/GoBackButton/GoBackButton";
import InfoBox from "../common/InfoBox/InfoBox";
import Spinner from "../common/Spinner/Spinner";
import styles from "./MovieDetails.module.scss";

const MovieDetails: React.FC = () => {
    const { id }: any = useParams();
    const dispatch = useDispatch();

    const {
        detailsMovieData,
        detailsMovieLoading,
        detailsMovieError,
    } = useSelector(selectDetailsMovie);

    useEffect(() => {
        dispatch(getDetailsMovieRequest(id));
    }, [dispatch, id]);

    const renderInfoBox = () => {
        if (detailsMovieLoading) return <Spinner />;
        if (detailsMovieError)
            return <p className={styles.message}>{detailsMovieError}</p>;

        return (
            <section className={styles.details}>
                <InfoBox {...detailsMovieData} />
            </section>
        );
    };

    return (
        <>
            <GoBackButton name="Back To List" />
            {renderInfoBox()}
        </>
    );
};

export default MovieDetails;
