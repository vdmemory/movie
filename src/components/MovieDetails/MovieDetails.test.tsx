import React from "react";
import { createRenderer } from "react-test-renderer/shallow";
import { useSelector } from "react-redux";

import MovieDetails from "./MovieDetails";
import GoBackButton from "../common/GoBackButton/GoBackButton";
import InfoBox from "../common/InfoBox/InfoBox";
import Spinner from "../common/Spinner/Spinner";
import { initialState } from "../../store/reducer";

jest.mock("react-router-dom", () => ({
    useParams: jest.fn().mockReturnValue({ id: "123" }),
}));

jest.mock("react-redux", () => ({
    ...jest.requireActual("react-redux"),
    useSelector: jest.fn(),
    useDispatch: jest.fn(),
}));

const setup = () => {
    const renderer = createRenderer();
    renderer.render(<MovieDetails />);
    const output = renderer.getRenderOutput();
    return { output, renderer };
};

const stateLoading = {
    ...initialState,
    loading: {
        ...initialState.loading,
        detailsMovie: true,
    }
};

const stateError = {
    ...initialState,
    error: {
        ...initialState.error,
        detailsMovie: "Error message",
    },
};

describe("MovieDetails Component", () => {
    
    describe("Render component and its components", () => {
        beforeEach(() => {
            // @ts-expect-error
            useSelector.mockImplementation((callback) => {
                return callback(initialState);
            });
        });
        afterEach(() => {
            // @ts-expect-error
            useSelector.mockClear();
        });
        test("Should renders correctly", () => {
            const { output } = setup();
            expect(output).toMatchSnapshot();
        });

        test("Should render GoBackButton component and check class", () => {
            const { output } = setup();
            const [component] = output.props.children;
            expect(component.type).toBe(GoBackButton);
        });

        test("Should render renderInfoBox, check class and its components", () => {
            const { output } = setup();
            const [, renderInfoBox] = output.props.children;
            expect(renderInfoBox.type).toBe("section");
            expect(renderInfoBox.props.className).toBe("details");
            expect(renderInfoBox.props.children.type).toBe(InfoBox);
        });
    });

    describe("Renderer of the component and its components, if the loading flag is true", () => {
        beforeEach(() => {
            // @ts-expect-error
            useSelector.mockImplementation((callback) => {
                return callback(stateLoading);
            });
        });
        afterEach(() => {
            // @ts-expect-error
            useSelector.mockClear();
        });
        test("Should renders correctly", () => {
            const { output } = setup();
            expect(output).toMatchSnapshot();
        });

        test("Should render Spinner component and check class", () => {
            const { output } = setup();
            const [, spinner] = output.props.children;
            expect(spinner.type).toBe(Spinner);
        });

    });

    describe("Renderer of the component and its components, if the error message", () => {
        beforeEach(() => {
            // @ts-expect-error
            useSelector.mockImplementation((callback) => {
                return callback(stateError);
            });
        });
        afterEach(() => {
            // @ts-expect-error
            useSelector.mockClear();
        });
        test("Should renders correctly", () => {
            const { output } = setup();
            expect(output).toMatchSnapshot();
        });

        test("Should render Error message and check class", () => {
            const { output } = setup();
            const [, error] = output.props.children;
            expect(error.type).toBe("p");
            expect(error.props.className).toBe("message");
            expect(error.props.children).toBe("Error message");
        });
    });
});