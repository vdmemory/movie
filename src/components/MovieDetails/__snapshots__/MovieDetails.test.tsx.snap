// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieDetails Component Render component and its components Should renders correctly 1`] = `
<React.Fragment>
  <GoBackButton
    name="Back To List"
  />
  <section
    className="details"
  >
    <InfoBox />
  </section>
</React.Fragment>
`;

exports[`MovieDetails Component Renderer of the component and its components, if the error message Should renders correctly 1`] = `
<React.Fragment>
  <GoBackButton
    name="Back To List"
  />
  <p
    className="message"
  >
    Error message
  </p>
</React.Fragment>
`;

exports[`MovieDetails Component Renderer of the component and its components, if the loading flag is true Should renders correctly 1`] = `
<React.Fragment>
  <GoBackButton
    name="Back To List"
  />
  <Spinner />
</React.Fragment>
`;
